### DISCUSSION TITLE
How to set up Machine and CPU names in the "GDB QEMU arm debugger"?

### DISCUSSION THREAD

**Discussion by eraus** [xpack-dev-tools/qemu-arm-xpack#24] [Q&A]
I was trying to create a minimum ARM Cortex-M4 project. I tried the template named "Hello World Arm Cortex-M C/C++ Project". I was able to build this one. Yet, I run into the problem of configuring the QEMU debugger. I want to use the new "GDB QEMU arm Debugger". Now, I was not sure about the choices for the following and used some guesses:

1. QEMU machine name: Used "netduinoplus2" (which is Cortex-M4 based).
2. QEMU CUP name: Used "Cortex-m4".

However, when running QEMU with the above choices, I see the following error in the Console:

> Invalid write at addr 0xFFFFFFFC, size 4, region '(null)', reason: rejected
> Invalid write at addr 0xFFFFFFE0, size 4, region '(null)', reason: rejected
> Invalid write at addr 0xFFFFFFC8, size 4, region '(null)', reason: rejected
> ...

My question is how to set up the above machine and CPU names so that I can run this template project?

Many thanks!

**Comment by ilg-ul** [comment]
> the template named "Hello World Arm Cortex-M C/C++ Project"

This is an old template, to demonstrate very simple projects, I would not recommend it for new projects.

> QEMU machine name: Used "netduinoplus2" (which is Cortex-M4 based)

This machine name is supported only by the deprecated `qemu-system-gnuarmeclipse`; please use the new `qemu-system-arm`. To get the name of the machines/cpus, see the release notes.

To start new unit test projects, use the "Hello World Arm QEMU xPack C/C++ Project" template.

For an example of a real life project with extensive multi-platform unit tests, see:

- <https://github.com/micro-os-plus/utils-lists-xpack/>

**Reply by eraus** [reply]
> This machine name is supported only by the deprecated qemu-system-gnuarmeclipse; please use the new qemu-system-arm. To get the name of the machines/cpus, see the release notes.

Yes, I was using the new `qemu-system-arm` and got the name of the machine/cpu from the release notes. (https://xpack.github.io/blog/2022/09/06/qemu-arm-v7-1-0-1-released/). I tried a few combinations using the "Hello World Arm Cortex-M C/C++ Project", and none was working. 

> To start new unit test projects, use the "Hello World Arm QEMU xPack C/C++ Project" template.

This was the first one I tried. It worked well. I have two issues with this template for my applications:

1. I need to unit test assembly code for Cortex-M3 and Cortex-M4. This template only works for Cortex-M0 or Cortex-M7. I am not sure how to add support for Cortex-M3 and M4. Even if I could, I would still run into the same problem as selecting the machine/cpu names. 
2. This template includes a lot of files for very simple projects. I prefer to have a simpler template like "Hello World Arm Cortex-M C/C++ Project" template if possible; this way, I can reduce the clutter in the project. 

So, the question is where I can find a tutorial of how to set up QEMU for the "Hello World Arm Cortex-M C/C++ Project" template using `qemu-system-arm`? Or, if I choose an arbitrary pair in the list of the release notes, how do I reconfigure the project's CPU to match this choice? Do I have to change the linker file?

I will take a look at the real-life project with extensive multi-platform unit tests at a later time. I don't have the needed background to understand it yet. My previous background is using Keil with multiple targets. This entire Eclipse/QEMU system is very new to me. 

Many thanks.

**Comment by ilg-ul** [comment]
> I tried a few combinations using the "Hello World Arm Cortex-M C/C++ Project", and none was working.

Well, things are generally not working by chance, you need to master all details.

> I need to unit test assembly code for Cortex-M3 and Cortex-M4. This template only works for Cortex-M0 or Cortex-M7. I am not sure how to add support for Cortex-M3 and M4. Even if I could, I would still run into the same problem as selecting the machine/cpu names.

Except floating point, the M[347] cores use the same set of instructions (Arm-v7m). I did a quick test with the M7F generated project, I changed the target to M4 and removed the hardware FP. The project compiled and ran fine.

> I prefer to have a simpler template like "Hello World Arm Cortex-M C/C++ Project" template if possible; 

That template does not generate functional projects, it requires you to add device specific files. Generally you have to match the device definitions, the memory map, the linker scripts, memory allocator, interrupt vectors, startup code, etc. All these details are already in the projects generated by the QEMU template, otherwise it wouldn't run.

> My previous background is using Keil with multiple targets. This entire Eclipse/QEMU system is very new to me.

Well, in this case you have to spend some time to figure out the details. 

BTW, the projects generated by the template are actually command line projects, with two different GUI coats (Eclipse and VS Code). The main advantage is that you can build and run them in an automated (scriptable) environment, like GitHub Actions. How did you run your Keil unit tests in a CI environment?

**Reply by eraus** [reply]
> Except floating point, the M[347] cores use the same set of instructions (Arm-v7m). I did a quick test with the M7F generated project, I changed the target to M4 and removed the hardware FP. The project compiled and ran fine.

Yes, they use the same non-floating-point instruction set. Yet, there are some differences with the number of cycles of some instructions that I was hoping to be accounted for in some of the assembly code. Anyway, it is great to hear that you can compile and run the project when changing to M4. It would be great if you could shed some light on how you made the changes.  

> That template does not generate functional projects, it requires you to add device specific files. Generally you have to match the device definitions, the memory map, the linker scripts, memory allocator, interrupt vectors, startup code, etc. All these details are already in the projects generated by the QEMU template, otherwise it wouldn't run.

Yes, I figured this is the case. Will try the modifications at a later time. 

> BTW, the projects generated by the template are actually command line projects, with two different GUI coats (Eclipse and VS Code). The main advantage is that you can build and run them in an automated (scriptable) environment, like GitHub Actions. How did you run your Keil unit tests in a CI environment?

Currently, we mainly have calculation-based assembly code, and the unit tests are run after the change of the code in Keil. We are looking at GitHub actions and GitLab CI/CD, but have not made decisions yet as this is not pressing. I'd like to hear your opinion if possible. 

Many thanks!

**Comment by ilg-ul** [comment]
> there are some differences with the number of cycles of some instructions that I was hoping to be accounted for in some of the assembly code.

How do you plan to do this? How do you count cycles? BTW, QEMU does not keep track of time, it compiles the target code into host instructions and executes them at its own pace.

> ... how you made the changes. 

Project properties, settings, c/c++. You need to learn Eclipse if you want to use it.

> the unit tests are run after the change of the code in Keil ...

Where does Eclipse fit into this picture? The Eclipse projects are not suitable for headless runs.

>  I'd like to hear your opinion if possible.

Yes, do a run with the utils-list tests and take a look how they are done. The xPacks tools around xpm were designed mainly to help with writing and running unit tests at GitHub Actions.

**Reply by eraus** [reply]
> How do you plan to do this? How do you count cycles? BTW, QEMU does not keep track of time, it compiles the target code into host instructions and executes them at its own pace.

We plan to use the CYCCNT register to read the cycle counts. Don't know if it works in QEMU yet. If not, this is not a big deal, we just run the real MCU periodically to see it. 

> Where does Eclipse fit into this picture? The Eclipse projects are not suitable for headless runs.

Currently, we run the unit tests in Keil manually as there is not a lot of testing. With Eclipse, we plan to do the same thing at the beginning. But eventually, we need to move towards the command line. That's why I am excited when you mentioned that the template project works on command line as well. 

I will play with your utils-list test and get familiar with it first. If I have follow-up questions, I will have a new discussion topic. 

BTW, I was reading a book titled "Modern Embedded Firmware Infrastructure: How to setup modern DevOps and continuous delivery infrastructure using Zephyr RTOS, CMock, Unity, RobotFramework and Renode". If you write a similar book about your xPacks, QEMU, Eclipse / command line, unit test automation, and continuous integration, that would be a quick sell. And a lot of people will benefit from it and your projects as well. I can do a free and careful proofreading for you:-)

Many thanks.

**Comment by ilg-ul** [comment]
> We plan to use the CYCCNT register to read the cycle counts.

Yeah, good luck with it. I guess it works, but except you'll read some increasing values, don't expect miracles.

> That's why I am excited when you mentioned that the template project works on command line as well.

Yes, it is a command line project, you can invoke it from the terminal:

- <https://github.com/micro-os-plus/hello-world-qemu-template-xpack/>

>  If you write a similar book about your xPacks, QEMU, Eclipse / command line, unit test automation, and continuous integration, that would be a quick sell

This is an interesting idea! I'm not sure I'm able to do it, but I will think about it. Thank you!

BTW, if you are familiar with unit tests, you might be interested in µTest++, the testing framework I use for my projects:

- <https://github.com/micro-os-plus/micro-test-plus-xpack/>

**Comment by ilg-ul** [comment]
> I will play with your utils-list test and get familiar with it first. If I have follow-up questions, I will have a new discussion topic.

For a real-life run you can take a look at:

- <https://github.com/micro-os-plus/utils-lists-xpack/actions/workflows/test-all.yml>

It runs the test on all imaginable platforms, and each test runs on lots of architectures, (native with different compilers and embedded for Cortex-M0/M7F, Cortex-A15/A72, RISC-V 32/64). I doubt you've ever seen a more comprehensive test.

**Reply by eraus** [reply]
> It runs the test on all imaginable platforms, and each test runs on lots of architectures, (native with different compilers and embedded for Cortex-M0/M7F, Cortex-A15/A72, RISC-V 32/64). I doubt you've ever seen a more comprehensive test.

Great! Will look at it.

**Comment by OmarBorroho** [comment]
Hello,

I have the same problem, I'm trying to debug an empty project using **qemu-system-arm**. I got to do it by using **qemu-system-gnuarmeclipse**. However, as the **gnuarmeclipse** is deprecated, I need to use the **QEMU arm** with a **cortex m4** or a **cortex m33**. I don't know if it is needed to change the setting of the project or the debug configuration.

As a start point, i'm trying to debug it in the e2Studio (an IDE build on eclipse). As you asked exectly the same question, I'm wondering if you got to get it.

Thanks!
